File: BOJ/9663/A.cpp
#include <iostream>
#define MAX 15
using namespace std;

int cols[MAX];
int N, total = 0;

bool check(int level)
{
    for (int i = 0; i < level; i++)
        if (cols[i] == cols[level] || abs(cols[level] - cols[i]) == level - i)
            return false;
    return true;
}

void bruteforce(int x)
{
    if (x == N)
        total++;
    else
    {
        for (int i = 0; i < N; i++)
        {
            cols[x] = i;
            if (check(x))
                bruteforce(x + 1);
        }
    }
}
int main()
{
    cin >> N;
    bruteforce(0);
    cout << total;
}

File: BOJ/9663/A.py
import sys

sys.stdin = open("D:\Projects\AA\BOJ\9663\input.txt")
input = sys.stdin.readline

MAX = 15
N = int(input().rstrip())
cols = [0 for _ in range(MAX+1)]

cnt = 0

# check if a queen can be placed in row y


def check(y):
    for i in range(0, y):
        if (cols[y] == cols[i]) or (cols[i] - i == cols[y] - y or cols[i] + i == cols[y] + y):
            return False
    return True


def bruteforce(size):
    global cnt
    if size == N:
        cnt = cnt + 1
    else:
        # iterate through all columns
        for i in range(N):
            cols[size] = i
            if(check(size)):
                bruteforce(size+1)


bruteforce(0)

print(cnt)
